Add IPFS Podcasting App

Cameron-IPFSPodcasting opened the pull request:
This is my first PR & app submission. Hope I did it right : )  It's an IPFS node/client that hosts podcast episodes using ipfspodcasting.net. Hope to have lightning integration in future releases.  - Thanks

# App Submission

### App name
IPFS Podcasting

### Version
0.5

### One line description of the app
_(max 50 characters)_

Crowd-host podcast episodes over IPFS

### Summary of the app
_(50 to 200 words)_

Turn your Umbrel into an IPFS node for self-hosting, crowd-hosting, and archiving of your favorite podcasts to the IPFS network.

IPFSPodcasting.net uses participating nodes to manage & track hashes of podcast episodes and generates IPFS/RSS feeds that load media from the IPFS network.

You only need to run the app to participate. You can optionally configure your email, and manage your node from the website. View the FAQ to learn more.

### Developer name
IPFSPodcasting.net

### Developer website
https://IPFSPodcasting.net/

### Source code link
_(Link to your app's source code repository.)_

https://github.com/Cameron-IPFSPodcasting/podcastnode

### Support link
_(Link to your Telegram support channel, GitHub issues/discussions, support portal, or any other place where users could contact you for support.)_

https://github.com/Cameron-IPFSPodcasting/podcastnode/issues

### Requires
- [ ] Bitcoin Core
- [ ] Electrum server
- [ ] LND

### 256x256 SVG icon
_(Submit an icon with no rounded corners as it will be dynamically rounded with CSS. GitHub doesn't allow uploading SVGs directly, so please upload your icon to an alternate service, like https://svgur.com, and paste the link below.)_

http://ipfspodcasting.net/ipfspod.svg
![ipfspod](https://user-images.githubusercontent.com/103131615/163884576-ac614b2c-4072-422f-a59d-11cc715b3984.svg)

### Gallery images
_(Upload 3 to 5 high-quality gallery images (1440x900px) of your app in PNG format, or just upload 3 to 5 screenshots of your app and we'll help you design the gallery images.)_

(screenshots... need your template with a gradient based on #f0b823 ?)

"Immutable podcast hosting using the IPFS network"
![1](https://user-images.githubusercontent.com/103131615/163884632-c59d85fc-e70a-4a86-b648-7dd9ab0658d7.png)

"Support your favorite podcasts with storage & bandwidth"
![2](https://user-images.githubusercontent.com/103131615/163884656-c2f98785-eed4-49c6-b59c-4048a291f372.png)

"Run a node to support crowd-hosted podcasting"
![3](https://user-images.githubusercontent.com/103131615/163884674-ae661458-733c-42dc-b1be-1a9258870c29.png)

### I have tested my app on:
- [x] [Umbrel dev environment](https://github.com/getumbrel/umbrel-dev)
- [x] [Umbrel OS on a Raspberry Pi 4](https://github.com/getumbrel/umbrel-os)
- [ ] [Custom Umbrel install on Linux](https://github.com/getumbrel/umbrel#-installation)

lukechilds commented on apps/ipfs-podcasting/docker-compose.yml:
```diff
@@ -0,0 +1,22 @@
+    stop_grace_period: 1m
+    user: "1000:1000"
+    ports:
+      - 4001:4001
+      - 5001:5001
+      - 8675:8675
```
What are each of these ports needed for?

Cameron-IPFSPodcasting replied:
4001 if for IPFS communication. Is it ok that this isn't over tor? If the umbrel is firewalled (like mine), they'll need to allow incoming/outgoing tcp/udp to port 4001

8675 is for the App UI (1st screenshot)

5001 is for the IPFS UI (2nd & 3rd screenshot)

lukechilds replied:
Having two different UIs on two different ports could be confusing. When the user clicks on the app icon we'll have to choose which UI to show. I see you have a link to the IPFS UI but that must hardcode the port. It would break if the port ever changes in the future. At some point in the future these ports will likely be assigned dynamically by Umbrel. Maybe we could run bot on the same port? The app UI could be server from `/admin` and `/*` could be proxied through to IPFS potentially?

You can checkout the Agora app where @yzernik used the same technique as an example.

Cameron-IPFSPodcasting replied:
Yes. It should work fine if they're behind NAT. I only mentioned port forwarding if they lock down all ports.

Without the IPFS interface (5001), you could run "ipfs swarm peers" in the docker shell to check/view peers.

When I first setup my umbrel I locked down everything except tor (9001), but found I needed to open 123 (NTP) and 80/443 to get bitcoin and lightning connections going.

lukechilds commented on scripts/configure:
```diff
@@ -262,6 +262,9 @@ APP_AGORA_SERVER_IP="10.21.21.88"
 APP_AGORA_FILEBROWSER_IP="10.21.21.89"
 APP_SPARKKIOSK_IP="10.21.21.90"
 APP_SPARKKIOSK_PORT="21214"
+APP_IPFSHOST_IP=`ip -o route get to 1.1.1.1 | sed -n 's/.*src \([0-9.]\+\).*/\1/p'`
```
Unfortunately we can't really do this. This file is just a short term hack for us to hardcode some settings that will in the future be generated on the fly by Umbrel and assigned to an app. We can't allow apps execute arbitrary commands here.

Does the app need to know the public IP to run? If it does, could this logic be done from inside the container?

lukechilds replied:
Hmmn, enaling CORS for everything could leak data from the IPFS API unless it implements CSRF protection (maybe it does I'm not familiar with IPFS).

Cameron-IPFSPodcasting replied:
I assumed the host IP was basically the same as "*" because everything is coming through the docker/host IP anyway (correct?). i.e. if they have access to the lan/host IP they'll have access to the IPFS API no matter how IPFS/CORS is configured.

lukechilds replied:
The attack vector would be CSRF from a malicious website so:

- Trick user to visit malicious website
- Malicious website runs JS that does fetch request to http://umbrel.local:5001/api?get-some-sensitive-ipfs-data
- Due to relaxed CORS policy, website can read sensitive repsonse

So I just drop a link to evil.com in Umbrel telegram and anyone that clicks the link while on their local network leaks IPFS data to me.

lukechilds commented:
Hey @Cameron-IPFSPodcasting, thanks for this app submission, really cool idea! So sorry for the delay in review!

Looks great, left some small comments.

Cameron-IPFSPodcasting commented:
Removed the host IP script.
Removed IPFS UI (port 5001) and locked down API access to (docker) localhost only.
Updated docker image to remove IPFS links from the App UI (port 8675)
Updated docker-compose.yml to use variables for ports 4001 (IPFS) & 8675 (app UI)

I'm sure users will want to see/use the IPFS UI, but it's not necessary for the app to function. Will research for a future release, or integrate into the app UI.

Will continue testing these changes on my Umbrel.

Let me know if there are any other issues.

Cameron

lukechilds commented:
Ok so I have the app running!
<img width="1218" alt="Screenshot 2022-05-12 at 21 11 42" src="https://user-images.githubusercontent.com/2123375/168095328-d65bcdd6-c1d2-410d-bbd8-3f0508146ccd.png">

Do I need to do something else to serve podcasting episodes or will this automatically mirror them?

Cameron-IPFSPodcasting commented:
Cool.
It will just sit there asking if there's anything to archive. If a new episode is published, you'll have a chance at hosting.

To "force" some work, you'd have to create an account, then "favorite" some feeds. If you "gold heart" a feed, it will start sending you episodes on the next request. Instructions/explanation is here... https://ipfspodcasting.net/Help/Favorites

You can see all the nodes here.
https://ipfspodcasting.net/PodSwarm

lukechilds commented:
Hmmn I toggled to gold but I don't appear to be downloading it. It would show up in the logs of the webui if I was right?

Cameron-IPFSPodcasting commented:
> I just created an account and it auto linked to my IPFS node, is that from the email I set in the app ui? If so maybe you should confirm that somehow? Otherwise couldn't anyone link their IPFS node to anyone elses account if they know their email address? Then it would leak all their liked podcasts to them.

If you changed your email on your node to someone else, you'd be giving control of the node to that person. They could store things on your node (so don't do that : )

lukechilds commented:
It looks like there is a CSRF vulnerability in the app. If a malicious website runs the following in their client side JavaScript:

```js
await fetch("http://umbrel.local:8675/", {
    "body": "email=hacker%40evil.com&storageMax=69",
    "method": "POST",
});
```

Then anyone who visits that website while on the same local network as their Umbrel will have the their storage set to 69GB and their node email set to hacker@evil.com. So the attacker would have control of their node.

lukechilds commented:
```
2022-05-12 14:56:36 : Response : {'download': 'https://mp3s.nashownotes.com/PC20-84-2022-05-06-Final.mp3', 'pin': '', 'filename': 'PC20-84-2022-05-06-Final.mp3', 'delete': '', 'message': 'Download, Pin, & Report hash. '}
2022-05-12 14:56:36 : Downloading https://mp3s.nashownotes.com/PC20-84-2022-05-06-Final.mp3
```

Woop! Download worked! Just took a minute to show. 🎉

lukechilds commented:
>It's a slow process. I only give you work when you ask, so updates on the website are only seen on the next request (every 10 minutes). There's no inbound connection to the node (other than the IPFS hosting part).

Got it!

>Will look into the CSRF issue with changing email.

Ok keep me posted!

Cameron-IPFSPodcasting commented:
Updated the docker image to add CSRF protection. It seems to be working. I'll also add an email notification via the website if/when I detect it has been changed on the IPFS node.

Here is a new screenshot for the app store without the IPFS UI links (should probably remove screenshots #2 & #3 since it's no longer in the app).
![1b](https://user-images.githubusercontent.com/103131615/168179516-25721c27-2d9d-4823-b609-f7e25ce67267.png)

Exciting stuff.

lukechilds commented:
Working great!

mayankchhabra commented:
Really cool app, Cameron! The overlap with v4v podcasting is exciting. Congrats on the [launch](https://twitter.com/umbrel/status/1525115653933649920)! 🎉

[Here](https://github.com/getumbrel/umbrel-apps-gallery/tree/master/ipfs-podcasting) are the gallery images and app icon for your app. If you ever want to change/update them, feel free to directly open a PR at [getumbrel/umbrel-apps-gallery](https://github.com/getumbrel/umbrel-apps-gallery). Cheers!

Cameron-IPFSPodcasting commented:
Thanks. I like the extra headphones on the screenshot!

Hoping to integrate LND to distribute v4v to the nodes if/when sats flow through the podcasts.

Listener Boost -> Podcaster -> IPFS Podcasting (with episode ID) -> the node who was hosting that episode when the boost arrived.

I'm sure it will be fractions of fractions, but hopefully the sats will add up.
